scripts/generate_git_info.py

```diff
@@ -0,0 +1,35 @@
+import os
+import subprocess
+
+# Path to the output header file
+header_file = "firmware/include/git_info.h"
+
+def get_git_info():
+    try:
+        # Get the current Git branch name
+        branch_name = subprocess.check_output(["git", "rev-parse", "--abbrev-ref", "HEAD"], stderr=subprocess.STDOUT)
+        branch_name = branch_name.strip().decode("utf-8")
+    except subprocess.CalledProcessError:
+        branch_name = "unknown"
+
+    try:
+        # Get the short commit ID
+        commit_id = subprocess.check_output(["git", "rev-parse", "--short", "HEAD"], stderr=subprocess.STDOUT)
+        commit_id = commit_id.strip().decode("utf-8")
+    except subprocess.CalledProcessError:
+        commit_id = "unknown"
+
+    return branch_name, commit_id
+
+# Get Git branch and commit ID
+branch_name, commit_id = get_git_info()
+
+# Ensure the include directory exists
+os.makedirs(os.path.dirname(header_file), exist_ok=True)
+
+# Write the branch name and commit ID to the header file
+with open(header_file, "w") as f:
```

Comment:
This is just a simple utility script, but there are a few things to note:
Generally it is considered bad practice to have top level statements like this, instead do this:
```python
def main():
    pass # your code here.

if __name__ == '__main__':
    main()
```

This allows re-using the script later on (so the code won't run when you import it, but can be called explicitly),
and if someone feels fancy, they can write a unit-test for it.

Apart from that, because it is a relative path you are using here, the script behaves different depending on how you start it.
e.g.:
```sh
py scripts/generate_git_info.py # Will work
cd scripts && py generate_git_info.py # Won't work.
```

What is common in cases like this is get the (absolute) path to the _current_ script:
`CURRENT_FILE = __file__`
Get the repo directory from that:
`REPO_DIR = os.path.dirname(os.path.dirname(CURRENT_FILE))`
and create the full path to the header:
`HEADER_FILE = os.path.join(REPO_DIR, 'firmware/include/git_info.h')`

Reply:
I know, but that's how it's recommended in the PIO action scripts.

https://docs.platformio.org/en/latest/scripting/actions.html

And these scripts are not supposed to be called manually. They are part of the build process.
I'll add some comments in the header of the file to make that more clear.

Reply:
I just tried it.
PIO does not run it as a program but probably loads it as a module.
If I put the code in main() it will not be executed.
Also __file__ is not defined in this environment.

Reply:
Right, but that example page also tells to use the `env.AddPreAction` or `env.AddPostAction` instead of blindly running the script.
That might behave better with dependency tracking and whatnot.

Reply:
You are right, I'll try to optimize that.

Reply:
I tried to move the code to a separate function `action(source, target, env)` and register it using

`Import("env")`
`env.AddPreAction("buildprog", action)`

but then it will not be executed during the build.
I also tried "$PROGPATH", "build" and "pre" as hook names, but that did not work either.

I'm not sure what else to try... :-(